### PR TITLE
Small fixes to enzyme PR

### PR DESCRIPTION
**Pull request by stelmo**
Small fixes to docs for smoment
Change gecko structure to include enzyme concentrations as variables

### CODE REVIEW

**Comment by cylon-x** [comment]
:green_circle: &nbsp;Commit c9b494b4 is formatted properly.

**Comment by cylon-x** [comment]
:green_circle: &nbsp;Commit c9766ea3 is formatted properly.

**Comment by cylon-x** [comment]
:red_square: &nbsp;Commit 36cdbcb5 requires formatting!

Required formatting changes summary:
```
 src/analysis/gecko.jl   | 7 +++++--
 src/analysis/smoment.jl | 3 ++-
 2 files changed, 7 insertions(+), 3 deletions(-)
```

**Comment by cylon-x** [comment]
:red_square: &nbsp;Commit f2d37f03 requires formatting!

Required formatting changes summary:
```
 src/analysis/gecko.jl   | 7 +++++--
 src/analysis/smoment.jl | 3 ++-
 2 files changed, 7 insertions(+), 3 deletions(-)
```

**Comment by cylon-x** [comment]
:red_square: &nbsp;Commit 64291b6a requires formatting!

Required formatting changes summary:
```
 src/analysis/gecko.jl   | 7 +++++--
 src/analysis/smoment.jl | 3 ++-
 2 files changed, 7 insertions(+), 3 deletions(-)
```

**Comment by cylon-x** [comment]
:red_square: &nbsp;Commit 64291b6a requires formatting!

Required formatting changes summary:
```
 src/analysis/gecko.jl   | 7 +++++--
 src/analysis/smoment.jl | 3 ++-
 2 files changed, 7 insertions(+), 3 deletions(-)
```

**Comment by cylon-x** [comment]
:red_square: &nbsp;Commit 54a6be83 requires formatting!

Required formatting changes summary:
```
 src/analysis/gecko.jl   | 9 ++++++---
 src/analysis/smoment.jl | 3 ++-
 2 files changed, 8 insertions(+), 4 deletions(-)
```

**Comment by cylon-x** [comment]
:red_square: &nbsp;Commit 09393470 requires formatting!

Required formatting changes summary:
```
 src/analysis/gecko.jl   | 9 ++++++---
 src/analysis/smoment.jl | 3 ++-
 2 files changed, 8 insertions(+), 4 deletions(-)
```

**Comment by cylon-x** [comment]
:red_square: &nbsp;Commit c240484a requires formatting!

Required formatting changes summary:
```
 src/analysis/gecko.jl                 |  4 ++--
 src/analysis/smoment.jl               |  3 ++-
 src/base/types/wrappers/GeckoModel.jl | 14 ++++++++------
 src/base/utils/gecko.jl               |  6 +++---
 4 files changed, 15 insertions(+), 12 deletions(-)
```

**Comment by cylon-x** [comment]
:red_square: &nbsp;Commit 5c6c4f92 requires formatting!

Required formatting changes summary:
```
 src/analysis/gecko.jl                 |  4 ++--
 src/analysis/smoment.jl               |  3 ++-
 src/base/types/wrappers/GeckoModel.jl | 21 +++++++++-------
 src/base/utils/enzymes.jl             |  2 +-
 src/base/utils/gecko.jl               |  9 ++++---
 test/analysis/gecko.jl                | 45 +++++++----------------------------
 6 files changed, 32 insertions(+), 52 deletions(-)
```

**Comment by cylon-x** [comment]
:red_square: &nbsp;Commit 6a539373 requires formatting!

Required formatting changes summary:
```
 src/base/utils/enzymes.jl |  2 +-
 test/analysis/gecko.jl    | 45 +++++++++------------------------------------
 2 files changed, 10 insertions(+), 37 deletions(-)
```

**Comment by cylon-x** [comment]
:red_square: &nbsp;Commit 4242a213 requires formatting!

Required formatting changes summary:
```
 src/analysis/gecko.jl | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)
```

**Comment by cylon-x** [comment]
:green_circle: &nbsp;Commit 0127dfc4 is formatted properly.

File: src/base/utils/gecko.jl [hunk 1]
```diff
@@ -70,17 +70,30 @@ _gecko_gene_product_coupling(model::GeckoModel) =
...
+Returns the fluxes (not gene product concentrations) of the model as a
+reaction-keyed dictionary, if solved. See [`protein_dict`](@ref) for a 
+function to get the gene product concentrations.
+"""
+flux_dict(model::GeckoModel, opt_model) =
+    is_solved(opt_model) ?
+    Dict(
+        fluxes(model) .=>
+            reaction_flux(model)' * value.(opt_model[:x])[1:n_reactions(model)],
+    ) : nothing
```
Comment (exaexa) [inline src/base/utils/gecko.jl:99]: this _must_ be derivable from `reaction_fluxes` and `fluxes`, having a custom magic overload for just GeckoModel is confusing. Do the `reaction_fluxes` right instead.

Comment (exaexa) [reply src/base/utils/gecko.jl:99]: (After the quick discussion: make a separate `enzyme_dict`, anyone can mix the dictionaries from `flux_dict` and `enzyme_dict` later themselves.)

**Review by exaexa — COMMENTED** [review]
looks good to go except a few pretty minor things

File: src/analysis/gecko.jl [hunk 2]
```diff
@@ -25,95 +24,71 @@ GECKO algorithm (see [`GeckoModel`](@ref) documentation for details).
   given gene product, again specified by string gene ID. By default, all gene
   products belong to group `"uncategorized"` which is the behavior of original
   GECKO.
-- `group_mass_limit` is a function that returns the maximum mass for a given
+- `gene_mass_group_bound` is a function that returns the maximum mass for a given
```
[the remark this answers is not in the document]
Comment (stelmo) [reply src/analysis/gecko.jl:27]: aahh, you are right, but I like the verbose one more :/

[the remark this answers is not in the document]
Comment (stelmo) [reply src/analysis/gecko.jl:27]: plus `gene_product...` is use elsehwere

File: src/base/types/wrappers/GeckoModel.jl [hunk 3]
```diff
@@ -163,14 +183,50 @@ function coupling_bounds(model::GeckoModel)
...
+"""
+balance(model::GeckoModel) =
+    [balance(model.inner); spzeros(length(model.coupling_row_gene_product))]
+
+"""
+    n_genes(model::GeckoModel)
+
+Return the number of genes that have enzymatic constraints associated with them.
+"""
+n_genes(model::GeckoModel) = length(genes(model))
```
[the remark this answers is not in the document]
Comment (stelmo) [reply src/base/types/wrappers/GeckoModel.jl:210]: made all of these functions more efficient

File: src/base/types/wrappers/GeckoModel.jl [hunk 4]
```diff
@@ -101,19 +108,29 @@ reactions(model::GeckoModel) =
...
 """
-    reactions(model::GeckoModel)
+    n_reactions(model::GeckoModel)
 
-Returns the number of all irreversible reactions in `model`.
+Returns the number of all irreversible reactions in `model` as well as the
+number of gene products that take part in enzymatic reactions.
 """
-n_reactions(model::GeckoModel) = length(model.columns)
+n_reactions(model::GeckoModel) = length(reactions(model))
```
[the remark this answers is not in the document]
Comment (stelmo) [reply src/base/types/wrappers/GeckoModel.jl:116]: why coupling length?

[the remark this answers is not in the document]
Comment (stelmo) [reply src/base/types/wrappers/GeckoModel.jl:116]: `n_reactions(model::GeckoModel) = length(model.columns)` should be fine

File: src/base/types/wrappers/GeckoModel.jl [hunk 5]
```diff
@@ -54,10 +51,11 @@ constraints are implemented using [`coupling`](@ref) and
 purely virtual and do not occur in [`metabolites`](@ref).
 """
 struct GeckoModel <: ModelWrapper
+    objective::SparseVec
```
[the remark this answers is not in the document]
Comment (stelmo) [reply src/base/types/wrappers/GeckoModel.jl:54]: so it is not a mutable struct, so changing the size will cause an error. However, if you only reconstruct it then you can't easily make the gene products objectives

File: src/base/types/wrappers/GeckoModel.jl [hunk 6]
```diff
@@ -163,14 +183,50 @@ function coupling_bounds(model::GeckoModel)
...
+Return the ids of all metabolites, both real and pseudo, for a [`GeckoModel`](@ref).
+"""
+metabolites(model::GeckoModel) = [metabolites(model.inner); genes(model) .* "#supply"]
+
+"""
+    n_metabolites(model::GeckoModel)
+
+Return the number of metabolites, both real and pseudo, for a [`GeckoModel`](@ref).
+"""
+n_metabolites(model::GeckoModel) = length(metabolites(model))
```
[the remark this answers is not in the document]
Comment (stelmo) [reply src/base/types/wrappers/GeckoModel.jl:232]: fixed :)

File: src/base/types/wrappers/GeckoModel.jl (hunk 5 above)
[the remark this answers is not in the document]
Comment (stelmo) [reply src/base/types/wrappers/GeckoModel.jl:54]: and keep it attached to the model

File: src/base/types/wrappers/GeckoModel.jl [hunk 7]
```diff
@@ -54,10 +51,11 @@ constraints are implemented using [`coupling`](@ref) and
 purely virtual and do not occur in [`metabolites`](@ref).
 """
 struct GeckoModel <: ModelWrapper
+    objective::SparseVec
     columns::Vector{_gecko_column}
     coupling_row_reaction::Vector{Int}
     coupling_row_gene_product::Vector{Tuple{Int,Tuple{Float64,Float64}}}
-    coupling_row_mass_group::Vector{Tuple{String,Float64}}
+    coupling_row_mass_group::Vector{Tuple{String,Vector{Int},Vector{Float64},Float64}}
```
[the remark this answers is not in the document]
Comment (stelmo) [reply src/base/types/wrappers/GeckoModel.jl:58]: yep!

File: src/base/utils/gecko.jl (hunk 1 above)
Comment (stelmo) [reply src/base/utils/gecko.jl:99]: In fact, I want to change `protein_dict` to `gene_product_dict` etc. this is more in keeping with how we named everything else in the pr

**Comment by cylon-x** [comment]
:green_circle: &nbsp;Commit efdd9fb1 is formatted properly.

**Comment by cylon-x** [comment]
:green_circle: &nbsp;Commit b027b2db is formatted properly.

**Comment by cylon-x** [comment]
:green_circle: &nbsp;Commit b027b2db is formatted properly.

**Pull request merged**